[CORE-13754] Make `disk_log_impl::adjacent_merge_compact()` pass by value

WillemKauf opened the pull request:
Fixes https://redpandadata.atlassian.net/browse/CORE-13754.

Passing the `segment_set` by reference could lead to a seg fault due to iterator invalidation by concurrent mutation of the underlying `_segs` container while it was being iterated over within `adjacent_merge_compact`.

This path is only taken when `log_compaction_use_sliding_window=false`.

Change the function signature to take `segment_set` by value, forcing the user to perform a copy of the segments container to avoid this issue.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`storage`: add check for `segment->is_closed()`

To prevent access of member variables within a closed `segment`.

This likely wasn't an issue but it is worth having the added check.

WillemKauf commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -567,7 +567,7 @@ ss::future<> disk_log_impl::adjacent_merge_compact(
         }
 
-        auto is_compactible_segment = !seg->has_appender()
+        auto is_compactible_segment = !seg->is_closed() && !seg->has_appender()
```
Bad bot

WillemKauf commented on src/v/storage/tests/log_truncate_test.cc:
```diff
@@ -565,8 +565,16 @@ TEST_F(storage_test_fixture, test_concurrent_truncate_and_compaction) {
+        }
+        segment_set segs(buf);
+        disk_log.adjacent_merge_compact(std::move(segs), compaction_cfg).get();
+        segment_set segs2(buf);
+        disk_log.adjacent_merge_compact(std::move(segs2), compaction_cfg).get();
+    }
```
Nah this is fine

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--81f5aa50-645c-4e1e-af74-6a21d879d394-->
<!--TEST RESULTS-->
<!--f0960f4b-4947-4913-a54f-708e9e78be63-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73509">build#73509</a></summary>


| test_class                          | test_method                              | test_arguments                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                      |
|:------------------------------------|:-----------------------------------------|:-------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "gzip"}   | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8c2-1bf9-4a7f-8edc-9f86f25fbaa2 | FLAKY         | 20/21    | upstream reliability is '79.00466562986003'. current run reliability is '95.23809523809523'. drift is -16.23343 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "gzip"}   | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8cc-0b6c-4fcb-9cdd-fd3845c3bc97 | FLAKY         | 20/21    | upstream reliability is '79.51807228915662'. current run reliability is '95.23809523809523'. drift is -15.72002 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "lz4"}    | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8c2-1bfa-47f9-aaa7-274b67b91bd4 | FLAKY         | 20/21    | upstream reliability is '80.19639934533551'. current run reliability is '95.23809523809523'. drift is -15.0417 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "lz4"}    | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8cc-0b6d-4639-9939-60bfbc2b2d3f | FLAKY         | 20/21    | upstream reliability is '80.69620253164557'. current run reliability is '95.23809523809523'. drift is -14.54189 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "snappy"} | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8c2-1bfc-4a36-894f-02b3c71f873b | FLAKY         | 20/21    | upstream reliability is '78.35218093699515'. current run reliability is '95.23809523809523'. drift is -16.88591 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "snappy"} | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8cc-0b6e-46b5-aff8-982dae1df8d6 | FLAKY         | 20/21    | upstream reliability is '78.90625'. current run reliability is '95.23809523809523'. drift is -16.33185 and the allowed drift is set to 50. The test should PASS          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "zstd"}   | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8c2-1bf2-4d15-914e-af00c5d42f9c | FLAKY         | 20/21    | upstream reliability is '73.74213836477988'. current run reliability is '95.23809523809523'. drift is -21.49596 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| JavaCompressionTest                 | test_upgrade_java_compression            | {"compression_type": "zstd"}   | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8cc-0b65-4159-8212-a2de9f936f36 | FLAKY         | 20/21    | upstream reliability is '73.74213836477988'. current run reliability is '95.23809523809523'. drift is -21.49596 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression                   |
| NodePostRestartProbeTest            | post_restart_probe_test                  | null                           | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8cc-0b6d-4639-9939-60bfbc2b2d3f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodePostRestartProbeTest&test_method=post_restart_probe_test                    |
| WriteCachingFailureInjectionE2ETest | test_crash_all                           | {"use_transactions": false}    | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8cc-0b6c-4fcb-9cdd-fd3845c3bc97 | FLAKY         | 17/21    | upstream reliability is '84.90127758420442'. current run reliability is '80.95238095238095'. drift is 3.9489 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                  |
| TxAtomicProduceConsumeTest          | test_basic_tx_consumer_transform_produce | {"with_failures": true}        | integration | https://buildkite.com/redpanda/redpanda/builds/73509#0199a8c2-1bf4-4a0d-8e5e-9757c846849d | FLAKY         | 20/21    | upstream reliability is '99.25558312655087'. current run reliability is '95.23809523809523'. drift is 4.01749 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73584">build#73584</a></summary>


| test_class                                      | test_method                       | test_arguments                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                  |
|:------------------------------------------------|:----------------------------------|:-----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite                                 | test_replica_pair_frequency       |                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/73584#0199ac2e-122d-48c8-815a-c41e8f101896 | FAIL          | 0/1      |                                                                                                                                                           |                                                                                                                                                               |
| MasterTestSuite                                 | test_replica_pair_frequency       |                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/73584#0199aca2-d612-4253-9cd7-9e2714deca91 | FAIL          | 0/1      |                                                                                                                                                           |                                                                                                                                                               |
| DatalakeTransactionTests                        | test_with_transactions            | {"cloud_storage_type": 1, "compaction": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73584#0199ac60-ee9a-488e-80ec-4e3c74b27b8a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeTransactionTests&test_method=test_with_transactions |
| update_properties_invalid_describe_configs_test | do_not_return_topic_config_no_mod |                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/73584#0199aca2-d612-4253-9cd7-9e2714deca91 | FAIL          | 0/1      |                                                                                                                                                           |                                                                                                                                                               |
</details>
<!--/body-->

oleiman commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1384,8 +1384,15 @@ ss::future<> disk_log_impl::do_compact(
     if (use_adjacent_merge) {
+        // Make a copy of `_segs` to avoid iterator invalidation in
+        // `adjacent_merge_compact`.
+        segment_set::underlying_t buf;
+        for (auto& seg : _segs) {
+            buf.push_back(seg);
+        }
+        segment_set segs(std::move(buf));
```
nitpick: should we just add a copy helper to `segment_set`?

WillemKauf replied:
Yeah after all the copy pasting I was tempted. But I feel like we don't really want to open that API up without other good reasons?

oleiman replied:
static member? if it feels wrong to you I'm not inclined to push it.

WillemKauf replied:
Added `segment_set::copy()`.

oleiman reviewed (commented):
looks reasonable. maybe push the loop into a helper function?

WillemKauf committed:
`storage`: add `segment_set::copy()`

Make the copy constructor `= default` but `private` in order
to provide access to a `copy()` helper.

WillemKauf committed:
`storage`: make `adjacent_merge_compact` pass by value

This could lead to a seg fault due to iterator invalidation by concurrent
mutation of the underlying `_segs` container while it was being iterated
over within `adjacent_merge_compact`.

Change the function signature to take `segment_set` by value, forcing
the user to perform a copy of the segments container to avoid this
issue.

oleiman reviewed (approved):
lgtm

dotnwat reviewed (approved):
lgtm.

> This could lead to a seg fault due to iterator invalidation by concurrent
mutation of the underlying `_segs` container while it was being iterated
over within `adjacent_merge_compact`.

can you please hypothesize about what a possible scenario is? for example, can a segment be added while adjacent_merge_compact is happening? what about removed? can adjacent_merge_compact invalidate while its iterating itself?

WillemKauf commented:
> for example, can a segment be added while adjacent_merge_compact is happening? what about removed?

any of the above are possible since no locks against segments or the container are being held in this place of iteration.

>can adjacent_merge_compact invalidate while its iterating itself?

No, this function won't mutate the container, only the segments within.

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27876-v25.2.x-574 remotes/upstream/v25.2.x
git cherry-pick -x b28c926208 002cda8ead 2ab0592b1c
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18246357556).

dotnwat commented:
> > for example, can a segment be added while adjacent_merge_compact is happening? what about removed?
> 
> any of the above are possible since no locks against segments or the container are being held in this place of iteration.
> 

Thanks. Please add that kind of detail to the commit messages in the future. It's likely top of mind at the exact moment you're creating the commit, but consumers of the commit message in the future will wonder what the races were
